site/src/components/pages/authWall.tsx

```diff
@@ -0,0 +1,42 @@
+import { NextPage } from "next";
+import { useRouter } from "next/router";
+import { VoidFunctionComponent } from "react";
+import { useUser } from "../../context/UserContext";
+import { SerializedUser } from "../../lib/model/user.model";
+
+export interface AuthWallPageContentProps {
+  user: SerializedUser;
+}
+
+export type AuthWallPageContent =
+  VoidFunctionComponent<AuthWallPageContentProps>;
+
+const AuthWallWrapper: VoidFunctionComponent<{
+  Content: AuthWallPageContent;
+}> = ({ Content }) => {
+  const router = useRouter();
+  const { user } = useUser();
+
+  if (user === "loading" || !process.browser) {
+    return null;
+  }
+
+  if (!user) {
+    void router.push("/");
+    return null;
+  }
+
+  return <Content user={user} />;
+};
+
+/**
+ * We don’t know if a user is logged in during server-side-rendering, but we want
+ * some pages to be available behind the auth wall. Once UserContext is populated
+ * with the result of /api/me, we want to redirect guest users from to /.
+ * use `withAuthWall` to enable this behavior.
+ *
+ * You can also get user object from Content props and thus avoid extra checks.
+ */
```

Comment:
We could find out if the user is logged in during SSR, since we have the cookie on the request.

We can do this in follow up since it is only an issue if the user attempts to visit a protected page directly. I've created [a task (internal)](https://app.asana.com/0/1201095311341924/1201715945634534/f).

Reply:
Yeah that might be useful for some pages. We’ll probably want to keep public content statically rendered to avoid SSR everywhere. This means that user loading state will be still there, we’ll just replace `Router.push()` with a redirect inside some `getServerSideProps`. I believe we can live with the current approach for quite long TBH.